### PR TITLE
Add UInt128 for u128

### PR DESCRIPTION
**Pull request by weikengchen**

### CODE REVIEW

**Comment by weikengchen** [comment]
On the way of fixing multiple errors. May take a longer time.

It seems that `u128` would be much more complicated because `addmany` uses u128 to compute the middle result, which would support `u64`but not `u128`. 

A separate implementation would be needed for `u128`, likely via [num-bigint](https://github.com/rust-num/num-bigint) or the BigInteger256.

**Comment by Pratyush** [comment]
Maybe you can use BigInteger256?

**Comment by weikengchen** [comment]
Passing. But a code review is needed. This implementation uses `num-bigint`, and it did not use `BigInteger256` due to the complexity of implementing it due to a more restrictive interface. However, adding another trait to the arkworks stack may always be a drawback.

**Comment by weikengchen** [comment]
I would merge this for now. But we likely need a revisit later (I will create an issue), very likely to replace with `BigInteger256`, since it is better to reduce the number of crates we depend on.

File: src/bits/uint.rs [hunk 1]
```diff
@@ -262,11 +272,9 @@ macro_rules! make_uint {
                     // Allocate each bit_gadget of the result
                     let mut coeff = F::one();
                     let mut i = 0;
-                    while max_value != 0 {
+                    while max_value != BigUint::zero() {
                         // Allocate the bit_gadget
-                        let b = AllocatedBit::new_witness(cs.clone(), || {
-                            result_value.map(|v| (v >> i) & 1 == 1).get()
-                        })?;
+                        let b = AllocatedBit::new_witness(cs.clone(), || result_value.clone().map(|v| (v >> i) & BigUint::one() == BigUint::one()).get())?;
```
Comment (Pratyush) [inline src/bits/uint.rs:277]: I think we can rewrite this more simply as 
```rust
let b = AllocatedBit::new_witness(cs.clone(), || result_value.as_ref().map(|v| (v >> i) == BigUint::one()).get())?;
```

File: src/bits/uint.rs [hunk 2]
```diff
@@ -246,7 +251,12 @@ macro_rules! make_uint {
                     }
 
                     // The value of the actual result is modulo 2^$size
-                    let modular_value = result_value.map(|v| v as $native);
+                    let modular_value = result_value.clone().map(|v|
```
Comment (Pratyush) [inline src/bits/uint.rs:254]: We can avoid the clone here by doing `as_ref()` instead

File: src/bits/uint.rs [hunk 3]
```diff
@@ -246,7 +251,12 @@ macro_rules! make_uint {
...
+                        {
+                            let modulus = BigUint::from(1u64) << ($size as u32);
```
Comment (Pratyush) [inline src/bits/uint.rs:256]: `BigUint::from(1u64)` -> `BigUint::one()`.

**Comment by Pratyush** [comment]
Sorry this review is kinda late lol

**Pull request merged**